### PR TITLE
[cloud usecase] zot cli should not download manifests when listing images

### PR DESCRIPTION
**Pull request by roxanaN**
Signed-off-by: Roxana Nemulescu <roxana.nemulescu@gmail.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this change require updates to the CNI daemonset config files to work?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cli/image_cmd_test.go [hunk 1]
```diff
@@ -502,8 +547,7 @@ func TestServerResponse(t *testing.T) {
 			str := space.ReplaceAllString(buff.String(), " ")
 			actual := strings.TrimSpace(str)
 			So(actual, ShouldContainSubstring, "IMAGE NAME TAG DIGEST SIZE")
-			So(actual, ShouldContainSubstring, "repo7 test:2.0 883fc0c5 15B")
-			So(actual, ShouldContainSubstring, "repo7 test:1.0 883fc0c5 15B")
+			So(actual, ShouldContainSubstring, "repo7 test:1.0 9beeea29 243B")
```
Comment (andaaron) [inline pkg/cli/image_cmd_test.go:550]: Why are the digests in the tests changing in this file?

File: pkg/cli/service.go [hunk 2]
```diff
@@ -368,59 +555,6 @@ func (service searchService) getCveByImage(ctx context.Context, config searchCon
 	rch <- stringResult{str, nil}
 }
 
-func groupCVEsBySeverity(cveList []cve) []cve {
```
Comment (andaaron) [inline pkg/cli/service.go:371]: We need to keep this implementation. with more severity levels.

File: pkg/cli/service.go [hunk 3]
```diff
@@ -474,10 +608,40 @@ func (service searchService) getFixedTagsForCVE(ctx context.Context, config sear
 	localWg.Wait()
 }
 
+func groupCVEsBySeverity(cveList []cve) []cve {
```
Comment (andaaron) [inline pkg/cli/service.go:611]: Replace this with the existing implementation on main branch.

File: pkg/extensions/search/schema.graphql [hunk 4]
```diff
@@ -67,10 +76,11 @@ type LayerInfo {
...
 type Query {
-  CVEListForImage(image: String!) :CVEResultForImage 
-  ImageListForCVE(id: String!) :[ImgResultForCVE]
-  ImageListWithCVEFixed(id: String!, image: String!) :ImgResultForFixedCVE
-  ImageListForDigest(id: String!) :[ImgResultForDigest]
-  ImageListWithLatestTag:[ImageInfo]
-  ExpandedRepoInfo(repo: String!):RepoInfo
+     CVEListForImage(image: String!) :CVEResultForImage
+     ImageListForCVE(id: String!) :[ImageInfo]
+     TagListForCve(id: String!, image: String!, getFixed: Boolean) :[ImageInfo]
```
Comment (andaaron) [inline pkg/extensions/search/schema.graphql:81]: `TagListForCve` is confusing, I don't know if I'm searching for images affected by a CVE or unaffected by a CVE.
Can we use `ImageListForFixedCVE`? Why the name change?

File: pkg/cli/cve_cmd_test.go [hunk 5]
```diff
@@ -379,7 +418,7 @@ func TestServerCVEResponse(t *testing.T) {
 		str := space.ReplaceAllString(buff.String(), " ")
 		str = strings.TrimSpace(str)
 		So(err, ShouldBeNil)
-		So(str, ShouldEqual, "IMAGE NAME TAG DIGEST SIZE zot-cve-test 0.0.1 63a795ca 75MB")
+		So(str, ShouldEqual, "")
```
Comment (andaaron) [inline pkg/cli/cve_cmd_test.go:421]: Why has this changed?

File: pkg/cli/image_cmd_test.go [hunk 6]
```diff
@@ -522,13 +569,12 @@ func TestServerResponse(t *testing.T) {
...
 			// Actual cli output should be something similar to (order of images may differ):
 			// IMAGE NAME    TAG       DIGEST    CONFIG    LAYERS    SIZE
-			// repo7         test:2.0  a0ca253b  b8781e88            15B
-			//                                             b8781e88  15B
-			// repo7         test:1.0  a0ca253b  b8781e88            15B
-			//                                             b8781e88  15B
+			// repo7         test:2.0  fdf5f251  dae17351            243B
+			//                                             2f2284ba  243B
+			// repo7         test:1.0  9beeea29  a09bf3b5            243B
+			//                                                		 243B
```
Comment (andaaron) [inline pkg/cli/image_cmd_test.go:575]: Replace tab with strings please.

File: pkg/cli/image_cmd_test.go [hunk 7]
```diff
@@ -522,13 +569,12 @@ func TestServerResponse(t *testing.T) {
...
 			So(actual, ShouldContainSubstring, "IMAGE NAME TAG DIGEST CONFIG LAYERS SIZE")
-			So(actual, ShouldContainSubstring, "repo7 test:2.0 883fc0c5 3a1d2d0c 15B b8781e88 15B")
-			So(actual, ShouldContainSubstring, "repo7 test:1.0 883fc0c5 3a1d2d0c 15B b8781e88 15B")
+			So(actual, ShouldContainSubstring, "repo7 test:1.0 9beeea29 a09bf3b5 243B 58f98639 243B")
```
Comment (andaaron) [inline pkg/cli/image_cmd_test.go:577]: I don't understand why this test returns other images compare to the one at line 535, the only difference is the `verbose` flag which formats the response differently?

File: pkg/cli/image_cmd_test.go [hunk 8]
```diff
@@ -570,7 +616,7 @@ func TestServerResponse(t *testing.T) {
 		})
 
 		Convey("Test image by digest", func() {
-			args := []string{"imagetest", "--digest", "883fc0c5"}
+			args := []string{"imagetest", "--digest", "9beeea29"}
```
Comment (andaaron) [inline pkg/cli/image_cmd_test.go:619]: Let's have a separate call on these tests. I don't understand why some digests need to be updated while others don't.

File: pkg/extensions/search/common/common.go [hunk 9]
```diff
@@ -25,7 +27,13 @@ type TagInfo struct {
 	Timestamp time.Time
 }
 
-func GetImageRepoPath(image string, storeController storage.StoreController) string {
+type ImageInfoByDigest struct {
```
Comment (andaaron) [inline pkg/extensions/search/common/common.go:30]: Where is this used?

Comment (roxanaN) [reply pkg/extensions/search/common/common.go:30]: This is not used. We use TagInfo instead

**Review by andaaron — APPROVED** [review]
Very good progress!
Now we just need to improve coverage. Are you sure we are testing both the old codepath requesting manifests one by one, and the new codepath using only graphql?

File: pkg/cli/cve_cmd.go [hunk 10]
```diff
@@ -131,8 +135,51 @@ type cveFlagVariables struct {
...
+	}
+
+	// nolint: gosec
+	resp, err := client.R().Head(extEndPoint)
+	if err != nil {
+		return false
+	}
+
+	if resp.StatusCode() == http.StatusOK {
+		searchEndPoint, err := combineServerAndEndpointURL(serverURL, constants.ExtSearchPrefix)
```
Comment (andaaron) [inline pkg/cli/cve_cmd.go:154]: `extentions` and `search` endpoints are needed for both the old logic and the new logic making a single graphql call.
We should check the new query to differentiate them by checking `ImageList` is available.

File: pkg/cli/service.go [hunk 11]
```diff
@@ -46,6 +57,196 @@ func NewSearchService() SearchService {
...
+func (service searchService) getImagesGQL(ctx context.Context, config searchConfig, username, password string,
+	imageName string,
+) (*imageListStructGQL, error) {
+	query := fmt.Sprintf(`{ImageList(repo: "%s") {`+`
+									RepoName Tag Digest ConfigDigest Size Layers {Size Digest}}
+							  }`,
+		imageName)
+	result := &imageListStructGQL{}
+
+	err := service.makeGraphQLQuery(ctx, config, username, password, query, result)
```
Comment (andaaron) [inline pkg/cli/service.go:69]: This logic seems to repeat itself in multiple functions, we could move it to a separate function, that should improve with coverage.

File: pkg/cli/client.go [hunk 12]
```diff
@@ -283,16 +284,12 @@ func (p *requestsPool) doJob(ctx context.Context, job *manifestJob) {
...
-	image.Tags = []tags{
-		{
-			Name:         job.tagName,
-			Digest:       digest,
-			Size:         size,
-			ConfigDigest: configDigest,
-			Layers:       layers,
-		},
-	}
+	image.RepoName = job.imageName
```
Comment (andaaron) [inline pkg/cli/client.go:287]: Logic to get manifests 1by1 is probably not getting triggered in the tests.

**Review by andaaron — APPROVED** [review]

**Comment by andaaron** [comment]
@roxanaN were read to merge this without other code changes (I rebased your branch a few time from GH UI and test are passing).

Before we do the merge (rebase your local branch first):
[  ] Update your commit message to reflect what your are changing as part of this PR, at the moment the commit message doesn't help at all
[  ] Make sure to sign your commit (after I rebased the signature was gone)

**Pull request merged**